### PR TITLE
Have combineReducers return same object if nothing changes

### LINKED ISSUE #853: Should `combineReducers` return the same object if nothing changes?
Right now, the reducer returned by `combineReducers` returns a new object regardless of whether or not the reducers it is combining make any changes. This means breaking reference equal every action, which limits its usefulness when using nested combinations unless you always select all the way down to the piece of state that maintains reference equality integrity. It'd be a little bit more slow to introduce an equality check, after the `mapValues` but it seems like it'd be better, yeah?

### PR DESCRIPTION
**Pull request by aaronjensen**
Fixes #853

### CODE REVIEW

File: test/utils/combineReducers.spec.js [hunk 1]
```diff
@@ -112,6 +112,17 @@ describe('Utils', () => {
...
+      const reducer = combineReducers({
+        child(state = {}) {
+          return state;
+        }
+      });
+
+      const initialState = reducer(undefined, '@@INIT');
+      expect(reducer(initialState, { type: 'FOO' })).toBe(initialState);
+    });
+
```
Comment (gaearon) [inline test/utils/combineReducers.spec.js]: I'd like to see more tests.
We need to check that:

* With several reducers, having just one return something different is enough to cause inequality
* With several reducers, having all return the same is enough to maintain equality

File: src/utils/combineReducers.js [hunk 2]
```diff
@@ -113,12 +113,16 @@ export default function combineReducers(reducers) {
       throw sanityError;
     }
 
+    var hasChanged = false;
+
```
Comment (gaearon) [inline src/utils/combineReducers.js]: Let's remove this newline; this declaration is directly related to the logic just below.

File: src/utils/combineReducers.js [hunk 3]
```diff
@@ -113,12 +113,16 @@ export default function combineReducers(reducers) {
...
     var finalState = mapValues(finalReducers, (reducer, key) => {
-      var newState = reducer(state[key], action);
+      var originalState = state[key];
+      var newState = reducer(originalState, action);
```
Comment (gaearon) [inline src/utils/combineReducers.js]: I think it's a little confusing now that we have many `*State` variables, almost all of which refer to slices of state, but some refer to the state object itself.

Let's rename `originalState` and `newState` to `previousStateByKey` and `nextStateByKey`?

**Comment by aaronjensen** [comment]
@gaearon I believe I've addressed your feedback, please let me know if there is anything else. Thanks!

**Comment by gaearon** [comment]
This is good to go, thanks.

**Pull request merged**